[Cocoa] NetworkSessionCocoa::createWebSocketTask creates up to 4 copies of the given URL request

whsieh opened the pull request:
#### 03f310261d6cc7d8e206a9885752800e654483b7
<pre>
[Cocoa] NetworkSessionCocoa::createWebSocketTask creates up to 4 copies of the given URL request
<a href="https://bugs.webkit.org/show_bug.cgi?id=245711">https://bugs.webkit.org/show_bug.cgi?id=245711</a>

Reviewed by Tim Horton.

Address a FIXME in `NetworkSessionCocoa::createWebSocketTask`; we clone the original URL request at
most once by instead lazily creating the mutable copy. This prevents us from needlessly copying the
request when two or more of the following are true:

- The protocol is non-null
- The request isn&apos;t app initiated
- Privacy proxy is disabled
- The main frame main resource was private relayed

No change in behavior.

* Source/WebKit/NetworkProcess/cocoa/NetworkSessionCocoa.mm:
(WebKit::NetworkSessionCocoa::createWebSocketTask):

Canonical link: <a href="https://commits.webkit.org/254911@main">https://commits.webkit.org/254911@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6a8dfc4bfd5b745531c7a608c66057fc4060df7e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/90612 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/35194 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/21227 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/99933 "Built successfully") | 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/33691 "Built successfully") | [  ~~🛠 mac-debug~~](https://ews-build.webkit.org/#/builders/71/builds/28855 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/82969 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/96354 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/96267 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/68/builds/26812 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/77446 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/26656 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/81590 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/81379 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/69681 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/34787 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/15417 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/32600 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/16396 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3431 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/36365 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/39323 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/38285 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/35483 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

hortont424 commented on Source/WebKit/NetworkProcess/cocoa/NetworkSessionCocoa.mm:
```diff
@@ -1679,46 +1679,40 @@ static void activateSessionCleanup(NetworkSessionCocoa& session, const NetworkSe
+    RetainPtr nsRequest = request.nsURLRequest(WebCore::HTTPBodyUpdatePolicy::DoNotUpdateHTTPBody);
+
+    auto makeRequestMutable = [&] {
+        if (auto *mutableRequest = dynamic_objc_cast<NSMutableURLRequest>(nsRequest.get()))
```
I wonder if isKindOfClass is a better plan than just keeping it in a different local...

whsieh replied:
Sounds good to me!

whsieh committed:
[Cocoa] NetworkSessionCocoa::createWebSocketTask creates up to 4 copies of the given URL request
https://bugs.webkit.org/show_bug.cgi?id=245711

Reviewed by Tim Horton.

Address a FIXME in `NetworkSessionCocoa::createWebSocketTask`; we clone the original URL request at
most once by instead lazily creating the mutable copy. This prevents us from needlessly copying the
request when two or more of the following are true:

- The protocol is non-null
- The request isn't app initiated
- Privacy proxy is disabled
- The main frame main resource was private relayed

No change in behavior.

* Source/WebKit/NetworkProcess/cocoa/NetworkSessionCocoa.mm:
(WebKit::NetworkSessionCocoa::createWebSocketTask):

Canonical link: https://commits.webkit.org/254911@main
